### PR TITLE
Refactor SubmissionConfigReader to use a map for the item process configurations based on entityType

### LINKED ISSUE #9217: When Solr cannot be found on startup, an odd NullPointerException may appear in logs *after* the important SolrServerException
**Describe the bug**
If you attempt to startup the DSpace backend *without* Solr (or if the backend cannot connect to Solr on startup) you may see an obscure/odd error like the following in your logs:
```
2023-11-30 21:01:31,229 ERROR unknown unknown org.springframework.boot.SpringApplication @ Application run failed
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'authenticationRestController': Unsatisfied dependency expressed through field 'converter'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'converterService': Unsatisfied dependency expressed through field 'converters'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'workflowItemConverter' defined in file [/dspace/webapps/server/WEB-INF/classes/org/dspace/app/rest/converter/WorkflowItemConverter.class]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.dspace.app.rest.converter.WorkflowItemConverter]: Constructor threw exception; nested exception is java.lang.NullPointerException
...
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'converterService': Unsatisfied dependency expressed through field 'converters'; 
...
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'workflowItemConverter' defined in file [/dspace/webapps/server/WEB-INF/classes/org/dspace/app/rest/converter/WorkflowItemConverter.class]: Instantiation of bean failed;
...
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.dspace.app.rest.converter.WorkflowItemConverter]: Constructor threw exception;
...
Caused by: java.lang.NullPointerException
  at org.dspace.app.rest.converter.AInprogressItemConverter.<init>(AInprogressItemConverter.java:63) ~[classes/:8.0-SNAPSHOT]
```

This obscure error is **misleading**, as the **root cause is that Solr is unavailable or the backend cannot communicate with Solr.**

**PRIOR TO** this exception, your logs will likely also display something like this:
```
2023-11-30 21:01:31,176 ERROR unknown unknown org.dspace.discovery.SolrSearchCore @ Error while initializing solr server http://dspacesolr:8983/solr/search
 org.apache.solr.client.solrj.SolrServerException: IOException occurred when talking to server at: http://dspacesolr:8983/solr/search
    at org.apache.solr.client.solrj.impl.HttpSolrClient.executeMethod(HttpSolrClient.java:695)
```
This prior exception is the **root cause** and it can "snowball" into the later obscure BeanInstantiationException / NullPointerException shown above.

**To Reproduce**
1. Stop Solr
2. Reboot the DSpace backend (i.e. Tomcat) (as this error only occurs on startup if Solr is not available)
3. Watch the `dspace.log` file and you'll see the above exceptions. First the useful `SolrServerException`, but then the misleading `NullPointerException` related to beans.

**Expected behavior**
Ideally, we'd find a way to avoid or not display the misleading exception as it's not useful & appears to imply a configuration or Spring Bean error.  The important `SolrServerException` information is slightly hidden just above that misleading exception.

At this time, however, I'm unsure of the cause of this misleading exception. So, this ticket just documents that it *exists* so that others who encounter it know to look for the `SolrServerException` just above it.

### LINKED ISSUE #9402: Default form is displayed for a Publication when I specified a proper submission-name
**Describe the bug**
Not all collections are being presented the correct configured form. 
It you have such a configuration like this in you item-submission.xml file:
```xml
<name-map collection-entity-type="Publication" submission-name="Publication" />
```
And have more than 10 collections with `dspace.entity.type = Publication` the 11th will not display the correct form. It will use the default one.

**To Reproduce**
1. You will need more than 10 collections with `dspace.entity.type = Publication` (or any other type)
2. Add more collections and associate the Publication type
3. Deposit one new item on a recent collection

**Expected behavior**
I'm expecting to have the correct form. But the default is shown.

**Related work**
- https://github.com/DSpace/DSpace/pull/9004
- https://github.com/DSpace/DSpace/pull/8855

### PR DESCRIPTION
**Pull request by toniprieto**
## References
* Fixes #9402
* Fixes #9217 (The strange NullPointerException will no longer occur. Now you get a clear error message that Solr is unavailable, e.g. `Failed to contact Solr at http://dspacesolr:8983/solr/search :  IOException occurred when talking to server at: http://dspacesolr:8983/solr/search`)
* Related to DSpace/dspace-angular#2914
* Port to 7x: #9487

## Description
This PR implements the changes described in [this comment](https://github.com/DSpace/dspace-angular/issues/2914#issuecomment-2050041593) to utilize a map for returning the item process submission form configured for collections with a specific entityType. The implementation follows a similar approach to #9259

## Instructions for Reviewers

List of changes in this PR:
* Refactor `SubmissionConfigReader` to use a new map for entityType-configured submission forms
* Add a new test to ensure that entityType-configured submission forms work correctly and take precedence over community-based configurations.
* Remove from default configuration the submissionconfig consumer used to reload the submission configuration after creating/modifying a collection. With the proposed changes, this reload should not be necessary. This change could be undone if it is useful for other use cases.

Test that the entityType configurations are working properly after creating o editing the entityType of a collection and that the issue #9402 is also resolved.

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.

### CODE REVIEW

**Review by paulo-graca — COMMENTED** [review]
Thank you @toniprieto for this contribution! With this PR, we don't need to reload (can also be configured to do it so) submission configuration every time we change a collection. It's a much better implementation. It also ensures that if we specify an handle, it will be given priority to that configuration.
I just have one request, if you within this PR, could remove unused collectionService code 
 - findAllCollectionsByEntityType.

File: dspace/config/dspace.cfg [hunk 1]
```diff
@@ -780,7 +780,7 @@ event.dispatcher.default.class = org.dspace.event.BasicDispatcher
 # Add rdf here, if you are using dspace-rdf to export your repository content as RDF.
 # Add iiif here, if you are using dspace-iiif.
 # Add orcidqueue here, if the integration with ORCID is configured and wish to enable the synchronization queue functionality
-event.dispatcher.default.consumers = versioning, discovery, eperson, submissionconfig, qaeventsdelete, ldnmessage
+event.dispatcher.default.consumers = versioning, discovery, eperson, qaeventsdelete, ldnmessage
```
Comment (tdonohue) [inline dspace/config/dspace.cfg:783]: @toniprieto : Should the "submissionconfig" consumer be entirely removed? (see lines 833-835) along with the `org.dspace.submit.consumer.SubmissionConfigConsumer` class?  Or is there still a use case for that SubmissionConfigConsumer?

If there is a use case for the SubmissionConfigConsumer, then we may want to enhance the comments on line 833 to explain when sites may want to enable it.

Comment (toniprieto) [reply dspace/config/dspace.cfg:783]: I had the same doubt. From the tests I've done, the consumer can be useful for reloading configuration changes in item-submission.xml without restarting Tomcat, mainly for adding new collection mappings. I ended up adding a comment in dspace.cfg, although I think it's a rare use case. If you think it's better to remove it entirely, let me know.

**Review by paulo-graca — COMMENTED** [review]
Thank you @toniprieto! All changes are looking perfect to me. It's a +1 on my behalf. I will try to test it in our on-premises setup.

**Comment by vins01-4science** [comment]
Thank you @toniprieto,

@tdonohue: I checked that this PR is solving the issue I was figuring out by starting the backend with the server-boot configuration (https://github.com/DSpace/DSpace/pull/9505#issuecomment-2090504847, https://github.com/DSpace/DSpace/issues/9476).
So if everything looks good for @tdonohue , I think that we can merge this PR in order to solve both issues :rocket: !

**Comment by tdonohue** [comment]
Thanks for the note @vins01-4science !  I'll test this today as well and, assuming it works for me, I'll get it merged immediately.  I'm very glad to hear this refactor should help solve issues that we were having in other PRs!

**Review by tdonohue — APPROVED** [review]
👍 Thanks @toniprieto !  This looks great.  Verified this fixes #9402.   I've also verified it fixes  #9217 as well (which is likely why this PR helps fix issues with other PRs, as I think #9217 was the root cause in those scenarios).

Merging immediately as this looks great and provides major benefits the the startup process as well as fixes SubmissionConfigReader bugs.

**Comment by tdonohue** [comment]
I've also updated the Submission Docs to remove mentions of having to restart Tomcat whenever a new Entity Collection is created (as this fixes that bug): https://wiki.lyrasis.org/display/DSDOC8x/Submission+User+Interface#SubmissionUserInterface-AssigningacustomSubmissionProcesstoaCollection

**Pull request merged**